File: observer/observer.py
import logging
from typing import List, Tuple
from model.event import T, Event
from model.state import State


class Observer(Event[T]):

    def __init__(self) -> None:
        pass

    async def process_all(self, state: State, events: List[Event[T]]) -> State:
        if len(events) == 0:
            return state
        current_state = state
        logging.info(
            f"[Observer]: received events from seq no {events[0].sequence_number} to {events[-1].sequence_number}: {events}")
        for event in events:
            try:
                (new_state, success) = await self.process(current_state, event)
                if not success:
                    return current_state
                current_state = new_state
            except Exception as err:
                logging.error(err)
                return current_state
        return current_state

    async def process(self, state: State, event: Event[T]) -> Tuple[State, bool]:
        pass

File: evm/src/observer/observer.ts
import { PrismaClientKnownRequestError } from "@prisma/client/runtime";
import { TypedEvent } from "../typechain/common";
import { State } from "../types";

export abstract class Observer {
  async processAll<T extends TypedEvent>(
    state: State,
    events: T[],
    _: number,
    endBlockNo: number
  ): Promise<State> {
    if (events.length == 0) {
      state[state.offsetField] = BigInt(endBlockNo);
      return { ...state };
    }
    let current_state = state;
    console.info(
      `[Observer]: received events from block no ${events[0].blockNumber} to ${
        events[events.length - 1].blockNumber
      }: ${events.map((e) => JSON.stringify(e as T))}`
    );
    for (const event of events) {
      try {
        let { state: new_state, success } = await this.process(
          current_state,
          event
        );
        if (!success) return current_state;
        current_state = new_state;
      } catch (e) {
        console.error("failed to process event:", e);
      }
    }
    return current_state;
  }

  abstract process<T extends TypedEvent>(
    state: State,
    event: T
  ): Promise<{ success: boolean; state: State }>;
}

export function processEvents<T extends TypedEvent>(
  state: State,
  events: T[],
  startBlock: number,
  endBlock: number,
  observer: Observer
): Promise<State> {
  return observer.processAll(state, events, startBlock, endBlock);
}

export function handleError(e: any) {
  let msg: string;
  if (e instanceof PrismaClientKnownRequestError) {
    msg = `${e.code}: ${JSON.stringify(e.meta)}`;
  } else {
    msg = e.toString();
  }
  console.error(msg);
}
